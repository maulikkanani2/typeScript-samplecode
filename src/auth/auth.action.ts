import { AuthError, CompleteNewPasswordParams, SignInParams, SignInResponse, User} from "./auth.type";
// ------------------------------------------
// Exported actions
// ------------------------------------------
export const SIGN_IN = 'SIGN_IN'
export const SIGN_IN_SUCCEEDED = 'SIGN_IN_SUCCEEDED'
export const SIGN_IN_FAILED = 'SIGN_IN_FAILED'

export const SIGN_OUT = 'SIGN_OUT'
export const SIGN_OUT_FAILED = 'SIGN_OUT_FAILED'
export const SIGN_OUT_SUCCEEDED = 'SIGN_OUT_SUCCEEDED'

export const COMPLETE_NEW_PASSWORD = 'COMPLETE_NEW_PASSWORD'
export const COMPLETE_NEW_PASSWORD_SUCCEEDED = 'COMPLETE_NEW_PASSWORD_SUCCEEDED'
export const COMPLETE_NEW_PASSWORD_FAILED = 'COMPLETE_NEW_PASSWORD_FAILED'

export const RETRIEVE_CURRENT_USER = 'RETRIEVE_CURRENT_USER'
export const RETRIEVE_CURRENT_USER_SUCCEEDED = 'RETRIEVE_CURRENT_USER_SUCCEEDED'
export const RETRIEVE_CURRENT_USER_FAILED = 'RETRIEVE_CURRENT_USER_FAILED'

export interface SignInAction {
    type: typeof SIGN_IN
    payload: SignInParams
}

export interface SignInSucceededAction {
    type: typeof SIGN_IN_SUCCEEDED
    payload: SignInResponse
}

export interface SignInFailedAction {
    type: typeof SIGN_IN_FAILED
    payload: AuthError
}

export interface SignOutAction {
    type: typeof SIGN_OUT
}

export interface SignOutActionSucceeded {
    type: typeof SIGN_OUT_SUCCEEDED
}

export interface SignOutActionFailed {
    type: typeof SIGN_OUT_FAILED
    payload: AuthError
}

export interface CompleteNewPasswordAction {
    type: typeof COMPLETE_NEW_PASSWORD
    payload: CompleteNewPasswordParams
}

export interface CompleteNewPasswordActionSucceeded {
    type: typeof COMPLETE_NEW_PASSWORD_SUCCEEDED
    payload: User
}

export interface CompleteNewPasswordActionFailed {
    type: typeof COMPLETE_NEW_PASSWORD_FAILED
    payload: AuthError
}

export interface RetrieveCurrentUserAction {
    type: typeof RETRIEVE_CURRENT_USER
}

export interface RetrieveCurrentUserActionSucceeded {
    type: typeof RETRIEVE_CURRENT_USER_SUCCEEDED
    payload: User
}

export interface RetrieveCurrentUserActionFailed {
    type: typeof RETRIEVE_CURRENT_USER_FAILED
    payload: AuthError
}

// export const signIn = ({email password}: SignInParams) => ({}
//     SIGN_IN,
//     authService.signIn
// )
//
// export const completeNewPassword = createAsyncAction(
//     COMPLETE_NEW_PASSWORD,
//     authService.completeNewPassword
// )
//
// export const signOut = createAsyncAction(
//     SIGN_OUT,
//     authService.signOut,
// )
//
// export const retrieveCurrentUser = createAsyncAction(
//     RETRIEVE_CURRENT_USER,
//     authService.retrieveCurrentUser,
// )

export type AuthActionTypes = SignInAction | SignInSucceededAction | SignInFailedAction |
    SignOutAction | SignOutActionSucceeded | SignOutActionFailed |
    CompleteNewPasswordAction | CompleteNewPasswordActionSucceeded | CompleteNewPasswordActionFailed |
    RetrieveCurrentUserAction | RetrieveCurrentUserActionSucceeded | RetrieveCurrentUserActionFailed
