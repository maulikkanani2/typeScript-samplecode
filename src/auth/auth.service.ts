import {
    AuthChallenge,
    CompleteNewPasswordParams,
    SignInParams,
    SignInResponse,
    User
} from "./auth.type";
import Auth, {CognitoUser} from "@aws-amplify/auth";


// This is authentication service file. It contains async function that handle authentication with remote services ( AWS Cognito )
// It should *NOT* leak remote service types and business logic back to the main app. Instead, it should convert the result into the types defined in auth.type.ts

// todo: handle error thrown by cognito
/**
 * Helper to convert cognito user to user
 * @param cognitoUser
 */
const convertToUser = (cognitoUser: CognitoUser): User => {
    return {
        userName: cognitoUser.getUsername(),
        cognitoUser,
    }
}

const tryGetChallenge = (cognitoUser: { challengeName?: string, challengeParam?: any }): AuthChallenge | undefined => {
    if (cognitoUser.challengeName) {
        return {
            challengeName: cognitoUser.challengeName,
            challengeParams: cognitoUser.challengeParam,
        }
    }
}
//
// const convertCognitoError = (cognitoError: any): Error => {
//     const {_type, message} = cognitoError
//     switch (_type) {
//         case 'UserNotFoundException':
//             return new AuthError(AuthErrorType.UserNotFound, message)
//         case 'InvalidPasswordException':
//             return new AuthError(AuthErrorType.InvalidPassword, message)
//         default:
//             return new Error(message)
//     }
// }

/**
 * Sign in a user. Return a user and a challenge if necessary
 * @param email
 * @param password
 */
export async function signIn({email, password}: SignInParams): Promise<SignInResponse> {
    const signInResponse = await Auth.signIn(email, password);
    const user = convertToUser(signInResponse)
    const challenge = tryGetChallenge(signInResponse)
    return {
        user, challenge
    }
}

/**
 * Complete new password required challenge
 * @param user
 * @param password
 * @param challengeParams
 */
export async function completeNewPassword({user, password, challengeParams}: CompleteNewPasswordParams): Promise<User> {
    const {requiredAttributes} = challengeParams
    const cognitoUser = await Auth.completeNewPassword(user.cognitoUser, password, requiredAttributes)
    return convertToUser(cognitoUser)
}

export async function signOut(): Promise<void> {
    await Auth.signOut();
}

export async function retrieveCurrentUser(): Promise<User> {
    const cognitoUser = await Auth.currentAuthenticatedUser();
    return convertToUser(cognitoUser)
}
