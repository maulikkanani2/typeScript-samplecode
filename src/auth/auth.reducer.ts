import * as authAction from "./auth.action";
import {AuthState} from "./auth.type";
import {
    COMPLETE_NEW_PASSWORD_FAILED,
    COMPLETE_NEW_PASSWORD_SUCCEEDED,
    CompleteNewPasswordActionSucceeded,
    RETRIEVE_CURRENT_USER_FAILED,
    RETRIEVE_CURRENT_USER_SUCCEEDED,
    RetrieveCurrentUserActionSucceeded,
    SIGN_IN_SUCCEEDED, SIGN_OUT_FAILED, SIGN_OUT_SUCCEEDED,
    SignInSucceededAction
} from "./auth.action";

const initialAuthState: AuthState = {
    pending: false
}

export default
(state: AuthState = initialAuthState, action: authAction.AuthActionTypes): AuthState => {
    switch (action.type) {
        case SIGN_IN_SUCCEEDED:
            const signInSucceededAction = action as SignInSucceededAction
            const { user, challenge} = signInSucceededAction.payload
            if (challenge && challenge.challengeName === 'NEW_PASSWORD_REQUIRED') {
                return {
                    ...state,
                    user,
                    challenge,
                    error: undefined,
                    pending: false,
                }
            }
            return {
                ...state,
                user,
                challenge: undefined,
                error: undefined,
                pending: false,
            }
        case COMPLETE_NEW_PASSWORD_SUCCEEDED:
            const completeNewPasswordActionSucceeded = action as CompleteNewPasswordActionSucceeded

            return {
                ...state,
                user: completeNewPasswordActionSucceeded.payload,
                challenge: undefined,
                error: undefined,
                pending: false,
            }
        case RETRIEVE_CURRENT_USER_SUCCEEDED:
            const retrieveCurrentUserActionSucceeded = action as RetrieveCurrentUserActionSucceeded
            return {
                ...state,
                user: retrieveCurrentUserActionSucceeded.payload,
                error: undefined,
                pending: false,
            }
        case SIGN_OUT_SUCCEEDED:
            return initialAuthState
        case SIGN_OUT_FAILED:
        case RETRIEVE_CURRENT_USER_FAILED:
        case COMPLETE_NEW_PASSWORD_FAILED:
            return {
                ...state,
                error: action.payload,
                pending: false,
            }

        default:
            return state
    }
}

