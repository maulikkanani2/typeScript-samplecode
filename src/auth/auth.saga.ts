import { takeLatest, call, put, all } from 'redux-saga/effects';
import * as authService from './auth.service'
import {
    COMPLETE_NEW_PASSWORD, COMPLETE_NEW_PASSWORD_FAILED, COMPLETE_NEW_PASSWORD_SUCCEEDED, CompleteNewPasswordAction,
    RETRIEVE_CURRENT_USER, RETRIEVE_CURRENT_USER_FAILED,
    RETRIEVE_CURRENT_USER_SUCCEEDED,
    SIGN_IN,
    SIGN_IN_FAILED,
    SIGN_IN_SUCCEEDED, SIGN_OUT, SIGN_OUT_FAILED, SIGN_OUT_SUCCEEDED,
    SignInAction
} from "./auth.action";

// watcher saga: watches for actions dispatched to the store, starts worker saga
export function* authSaga() {
    yield all([
        takeLatest(SIGN_IN, signInWorkerSaga),
        takeLatest(COMPLETE_NEW_PASSWORD, completeNewPasswordWorkerSaga),
        takeLatest(RETRIEVE_CURRENT_USER, retrieveCurrentUserWorkerSaga),
        takeLatest(SIGN_OUT, signOutWorkerSaga),
    ])

}

function* signInWorkerSaga(action: SignInAction) {
    try {
        const signInResponse = yield call(authService.signIn, action.payload);

        yield put({ type: SIGN_IN_SUCCEEDED, payload: signInResponse });

    } catch (error) {
        // dispatch a failure action to the store with the error
        yield put({ type: SIGN_IN_FAILED, payload: error });
    }
}

function* completeNewPasswordWorkerSaga(action: CompleteNewPasswordAction) {
    try {
        const response = yield call(authService.completeNewPassword, action.payload);

        yield put({ type: COMPLETE_NEW_PASSWORD_SUCCEEDED, payload: response });

    } catch (error) {
        // dispatch a failure action to the store with the error
        yield put({ type: COMPLETE_NEW_PASSWORD_FAILED, payload: error });
    }
}

function* retrieveCurrentUserWorkerSaga() {
    try {
        const response = yield call(authService.retrieveCurrentUser);

        yield put({ type: RETRIEVE_CURRENT_USER_SUCCEEDED, payload: response });

    } catch (error) {
        // dispatch a failure action to the store with the error
        yield put({ type: RETRIEVE_CURRENT_USER_FAILED, error });
    }
}

function* signOutWorkerSaga() {
    try {
        yield call(authService.signOut);

        yield put({ type: SIGN_OUT_SUCCEEDED });

    } catch (error) {
        // dispatch a failure action to the store with the error
        yield put({ type: SIGN_OUT_FAILED, error });
    }
}
