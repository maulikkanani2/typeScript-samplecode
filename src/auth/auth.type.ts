import {CognitoUser} from "@aws-amplify/auth";

export interface SignInParams {
    email: string
    password: string
}

interface ChallengeParams {
    requiredAttributes: any
}
export interface CompleteNewPasswordParams {
    user: User,
    password: string,
    challengeParams: ChallengeParams,
}
export interface AuthChallenge {
    challengeName: string,
    challengeParams: any,
}

export interface User {
    userName: string,
    cognitoUser: CognitoUser,
}

export interface SignInResponse {
    user: User,
    challenge?: AuthChallenge,
}

export interface Error {

}

export interface AuthState {
    readonly user?: User,
    readonly challenge?: AuthChallenge,
    readonly error?: AuthError
    readonly pending: boolean
}

export enum AuthErrorType {
    UserNotFound = 'UserNotFound',
    InvalidPassword = 'InvalidPassword',
    InvalidLogin = 'InvalidLogin'
}

export class AuthError extends Error {
    public type: AuthErrorType
    public message: string

    public constructor(type: AuthErrorType, message: string) {
        super(message);
        this.type = type;
        this.message = message;
    }
}
