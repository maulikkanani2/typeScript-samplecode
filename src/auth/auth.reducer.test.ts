import { SignInResponse , AuthChallenge} from "./auth.type";
import * as authActions from "./auth.action";
import authReducers from "./auth.reducer";
describe("auth.reducer", () => {

  describe("undefined action", () => {
    it("should return the default state", () => {
      const action = {} as any;      
      expect(authReducers(undefined, action)).toMatchObject({pending: false});
    });
  });

  describe("SIGN_IN_SUCCEEDED action", () => {
    it("it should return sign In success state with user", async function() {     
        const user = { userName: "sogisiyi@freetmail.in", cognitoUser: null };
        const result: SignInResponse = { user: user,  challenge: undefined, };
        const action: authActions.SignInSucceededAction = {
          type: authActions.SIGN_IN_SUCCEEDED,  
          payload: result,
        };              
        expect(authReducers({ pending: true}, action)).toMatchSnapshot();     
    });
    it("it should return new password In success  state with user", async function() {     
        const user = { userName: "sogisiyi@freetmail.in", cognitoUser: null };
        const Challenge : AuthChallenge = { challengeName :'NEW_PASSWORD_REQUIRED' , challengeParams:{username:user.userName} }
        const result: SignInResponse = { user: user,  challenge: Challenge };
        const action: authActions.SignInSucceededAction = {type: authActions.SIGN_IN_SUCCEEDED, payload: result };              
        expect(authReducers({ pending: true}, action)).toMatchSnapshot();     
    });
  });

  describe("RETRIEVE_CURRENT_USER_SUCCEEDED action", () => {
    it("it should return from retive current user", async function() {     
        const user = { userName: "sogisiyi@freetmail.in", cognitoUser: null };
        const action: authActions.RetrieveCurrentUserActionSucceeded = {
          type: authActions.RETRIEVE_CURRENT_USER_SUCCEEDED,          
          payload: user,
        };                    
        expect(authReducers({pending : true}, action)).toMatchSnapshot();      
    });
  });

  describe("COMPLETE_NEW_PASSWORD_SUCCEEDED action", () => {
    it("it should return new password with user", async function() {   
        const user = { userName: "sogisiyi@freetmail.in", cognitoUser: null };
        const action: authActions.CompleteNewPasswordActionSucceeded = {
          type: authActions.COMPLETE_NEW_PASSWORD_SUCCEEDED,
          payload: user,
        };
        expect(authReducers({pending : true}, action)).toMatchSnapshot();     
    });
  });

  describe("SIGN_OUT_SUCCEEDED action", () => {
    it("it should return from sign out user action", () => {
      let action: authActions.SignOutActionSucceeded = {
        type: authActions.SIGN_OUT_SUCCEEDED,
      };      
      return expect(authReducers({pending : true}, action)).toEqual({pending : false});
    });
  });

  describe("COMPLETE_NEW_PASSWORD_FAILED action", () => {
    it("it should return faild user password deatils", () => {      
      let action: authActions.CompleteNewPasswordActionFailed = {
        type: authActions.COMPLETE_NEW_PASSWORD_FAILED,
        payload: null,
      };
      return expect(authReducers({pending : true}, action)).toEqual({ pending: false, error: null});
    });
  });

  describe("SIGN_OUT_FAILED action", () => {
    it("it should return sign out fail reason ", () => {   
      let action: authActions.SignOutActionFailed = {
        type: authActions.SIGN_OUT_FAILED,
        payload: null,
      };
      return expect(authReducers({pending : true}, action)).toEqual({pending: false , error : null});
    });
  });

  describe("RETRIEVE_CURRENT_USER_FAILED action", () => {
    it("it should return next state with user", () => {
      const user = { userName: "sogisiyi@freetmail.in", cognitoUser: null };      
      let action: authActions.RetrieveCurrentUserActionFailed = {
        type: authActions.RETRIEVE_CURRENT_USER_FAILED,
        payload: undefined,
      };
      return expect(authReducers({pending: false, user: user}, action)).toMatchSnapshot();
    });
  });
});
