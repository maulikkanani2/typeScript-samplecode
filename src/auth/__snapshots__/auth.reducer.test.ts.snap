// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auth.reducer COMPLETE_NEW_PASSWORD_SUCCEEDED action it should return new password with user 1`] = `
Object {
  "challenge": undefined,
  "error": undefined,
  "pending": false,
  "user": Object {
    "cognitoUser": null,
    "userName": "sogisiyi@freetmail.in",
  },
}
`;

exports[`auth.reducer RETRIEVE_CURRENT_USER_FAILED action it should return next state with user 1`] = `
Object {
  "error": undefined,
  "pending": false,
  "user": Object {
    "cognitoUser": null,
    "userName": "sogisiyi@freetmail.in",
  },
}
`;

exports[`auth.reducer RETRIEVE_CURRENT_USER_SUCCEEDED action it should return from retive current user 1`] = `
Object {
  "error": undefined,
  "pending": false,
  "user": Object {
    "cognitoUser": null,
    "userName": "sogisiyi@freetmail.in",
  },
}
`;

exports[`auth.reducer SIGN_IN_SUCCEEDED action it should return new password In success  state with user 1`] = `
Object {
  "challenge": Object {
    "challengeName": "NEW_PASSWORD_REQUIRED",
    "challengeParams": Object {
      "username": "sogisiyi@freetmail.in",
    },
  },
  "error": undefined,
  "pending": false,
  "user": Object {
    "cognitoUser": null,
    "userName": "sogisiyi@freetmail.in",
  },
}
`;

exports[`auth.reducer SIGN_IN_SUCCEEDED action it should return sign In success state with user 1`] = `
Object {
  "challenge": undefined,
  "error": undefined,
  "pending": false,
  "user": Object {
    "cognitoUser": null,
    "userName": "sogisiyi@freetmail.in",
  },
}
`;
